### PR TITLE
Generalised rush larsen

### PR DESCRIPTION
**Pull request by MauriceHendrix**
## Description

- Generalised rush larsen First Order models and Opt versions with partial eval of Generalised rush larsen First Order (lookup tables not implemented)

- I had to change partial eval to only use piecewise_fold if there actually are any piecewises in the equation, as one of the Opt models got stuck (or it took more time than I have patience for) in the piecewise fold. This resulted in quite a few of the other reference models changing slightly.

## Motivation and Context

The penultemate type of model on the list in the github issue, and needed to move towards pycml replacement.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] I have updated all documentation necessary.
- [ ] I have checked spelling in (new) comments.

## Testing
- [X] Testing is done automatically and codecov shows test coverage
- [ ] This cannot be tested automatically 

All models of all types have been (re-) tested in chaste

### CODE REVIEW

File: chaste_codegen/__init__.py [hunk 1]
```diff
@@ -6,6 +6,8 @@
 from cellmlmanip.transpiler import Transpiler
 
 from chaste_codegen.backward_euler_model import BackwardEulerModel  # noqa
+from chaste_codegen.generalised_rush_larsen_1_model import GeneralisedRushLarsenModelFirstOrder  # noqa
```
Comment (jonc125) [inline chaste_codegen/__init__.py:9]: Would `GeneralisedRushLarsenFirstOrderModel` be more consistent?

File: chaste_codegen/generalised_rush_larsen_1_model.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        super().__init__(model, file_name, **kwargs)
+        self._hpp_template = 'generalised _rush_larsen_model_1.hpp'
+        self._cpp_template = 'generalised_rush_larsen_model_1.cpp'
+        self._jacobian_equations, self._jacobian_matrix = self._get_jacobian()
+        self._vars_for_template['jacobian_equations'], self._vars_for_template['jacobian_entries'] = \
+            format_jacobian(self._jacobian_equations, self._jacobian_matrix, self._printer, skip_0_entries=False)
+        self._map_state_vars_and_eqs()
+
+    def _get_jacobian(self):
+        """Retreive jacobian matrix"""
```
Comment (jonc125) [inline chaste_codegen/generalised_rush_larsen_1_model.py:21]: Retreive -> Retrieve

File: chaste_codegen/tests/console_script_help.txt [hunk 3]
```diff
@@ -0,0 +1,43 @@
+usage: chaste_codegen [-h] [--version] [-t TYPE] [-o OUTFILE] [-j]
```
Comment (jonc125) [inline chaste_codegen/tests/console_script_help.txt:1]: I thought this file already existed?

File: chaste_codegen/tests/console_script_help.txt [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+positional arguments:
+  cellml_file           The cellml file or URI to convert to chaste code
+
+optional arguments:
+  -h, --help            show this help message and exit
+  --version             show program's version number and exit
+  -t TYPE, --translate-type TYPE
+                        the type of code to output [default: Chaste]. Choices:
+                        ['Chaste', 'ChasteOpt', 'CVODE', 'BackwardsEuler',
```
Comment (jonc125) [inline chaste_codegen/tests/console_script_help.txt:15]: The generalised case is missing here?

File: chaste_codegen/tests/hilgemann_noble_chaste_codegen.txt [hunk 5]
```diff
@@ -0,0 +1,220 @@
+var_x0 = 0.074867781645490933 * var_chaste_interface__membrane__V;
```
Comment (jonc125) [inline chaste_codegen/tests/hilgemann_noble_chaste_codegen.txt:1]: Added in error?

File: chaste_codegen/tests/hilgemann_noble_model_pycml.txt [hunk 6]
```diff
@@ -0,0 +1,404 @@
+var_chaste_interface__membrane__V = (mSetVoltageDerivativeToZero ? this->mFixedVoltage : NV_Ith_S(rY, 0));
```
Comment (jonc125) [inline chaste_codegen/tests/hilgemann_noble_model_pycml.txt:1]: Added in error?

File: chaste_codegen/tests/test_all_models.py [hunk 7]
```diff
@@ -82,8 +84,28 @@ def test_RL_all_models(tmp_path, model, request):
...
+    if request.config.option.markexpr != 'all_models':
+        pytest.skip('Skip if not explicitly set to run all_models with -m all_models')
+    from chaste_codegen.tests.test_codegen import test_GRL1
+    test_GRL1(tmp_path, model)
+
+
+@pytest.mark.all_models
+@pytest.mark.parametrize(('model'), chaste_all_GRL1Opt)
+def test_GRL1Opt_all_models(tmp_path, model, request):
+    """ Check generation of Generalised Rush Larsen First Opt order models against reference"""
```
Comment (jonc125) [inline chaste_codegen/tests/test_all_models.py:107]: First Opt order -> First order optimised

File: chaste_codegen/tests/hilgemann_noble_chaste_codegen.txt (hunk 5 above)
Comment (MauriceHendrix) [reply chaste_codegen/tests/hilgemann_noble_chaste_codegen.txt:1]: hhm not sure why they've come back

File: chaste_codegen/tests/console_script_help.txt (hunk 3 above)
Comment (MauriceHendrix) [reply chaste_codegen/tests/console_script_help.txt:1]: it did

**Pull request merged**